File: remoteassets.rb
#!/usr/bin/env ruby
#
# Remote Assets
# http://cixtor.com/
# https://github.com/cixtor/mamutools
#
# List all the resources (assets) loaded in a specific page on a website.
#
# A digital asset is any item of text or media that has been formatted into a
# binary source that includes the right to use it. A digital file without the
# right to use it is not an asset. Digital assets are categorised in three major
# groups which may be defined as textual content (digital assets), images (media
# assets) and multimedia (media assets).
#
require 'rubygems'
require 'optparse'
require 'open-uri'
require 'uri'

config = Hash.new
optparse = OptionParser.new do |opt|
    opt.banner = "Usage: ruby #{opt.program_name} [OPTIONS]"
    opt.version = 1.0

    opt.on('-u', '--url args', 'Specify the remote location to scan.'){ |args| config[:target] = args }
    opt.on('-t', '--type args', 'Specify the filetype to look in the remote location specified.'){ |args| config[:filetype] = args }
    opt.on('-r', '--resource [images,javascript,styles]', 'Specify a group of filestypes to include in the scanning, for example, if you choose IMAGES it will search for the most common extensions: gif, jpg, png.'){ |args| config[:resource] = args }
    opt.on('-b', '--scanby [tag,extension]', 'Specify the type of scanning, you can choose between a scan using the tag architecture or using the extension of each file'){ |args| config[:scanby] = args }
end

def random_agent
    user_agents = [
        'Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; .NET CLR 1.1.4322; .NET CLR 2.0.50727; .NET CLR 3.0.04506.30)',
        'Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; .NET CLR 1.1.4322)',
        'Googlebot/2.1 (http://www.googlebot.com/bot.html)',
        'Opera/9.20 (Windows NT 6.0; U; en)',
        'Mozilla/5.0 (X11; U; Linux i686; en-US; rv:1.8.1.1) Gecko/20061205 Iceweasel/2.0.0.1 (Debian-2.0.0.1+dfsg-2)',
        'Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0; FDM; .NET CLR 2.0.50727; InfoPath.2; .NET CLR 1.1.4322)',
        'Opera/10.00 (X11; Linux i686; U; en) Presto/2.2.0',
        'Mozilla/5.0 (Windows; U; Windows NT 6.0; he-IL) AppleWebKit/528.16 (KHTML, like Gecko) Version/4.0 Safari/528.16',
        'Mozilla/5.0 (compatible; Yahoo! Slurp/3.0; http://help.yahoo.com/help/us/ysearch/slurp)',
        'Mozilla/5.0 (X11; U; Linux x86_64; en-US; rv:1.9.2.13) Gecko/20101209 Firefox/3.6.13',
        'Mozilla/4.0 (compatible; MSIE 9.0; Windows NT 5.1; Trident/5.0)',
        'Mozilla/5.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 1.1.4322; .NET CLR 2.0.50727)',
        'Mozilla/4.0 (compatible; MSIE 7.0b; Windows NT 6.0)',
        'Mozilla/4.0 (compatible; MSIE 6.0b; Windows 98)',
        'Mozilla/5.0 (Windows; U; Windows NT 6.1; ru; rv:1.9.2.3) Gecko/20100401 Firefox/4.0 (.NET CLR 3.5.30729)',
        'Mozilla/5.0 (X11; U; Linux x86_64; en-US; rv:1.9.2.8) Gecko/20100804 Gentoo Firefox/3.6.8',
        'Mozilla/5.0 (X11; U; Linux x86_64; en-US; rv:1.9.2.7) Gecko/20100809 Fedora/3.6.7-1.fc14 Firefox/3.6.7',
        'Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)',
        'Mozilla/5.0 (compatible; Yahoo! Slurp; http://help.yahoo.com/help/us/ysearch/slurp)',
        'YahooSeeker/1.2 (compatible; Mozilla 4.0; MSIE 5.5; yahooseeker at yahoo-inc dot com ; http://help.yahoo.com/help/us/shop/merchant/)'
    ]
    user_agents[rand(user_agents.size)]
end

begin
    optparse.parse!(ARGV)
    if config[:target].nil? then
        raise OptionParser::MissingArgument
    else
        config[:target] = "http://#{config[:target]}" if !config[:target].match(/^(http|https):\/\/.*/)
        config[:hostname] = URI.parse(config[:target]).host
        config[:scanby] = config[:scanby].nil? ? 'tag' : config[:scanby]

        filetypes = Array.new
        filetypes << config[:filetype] if !config[:filetype].nil?
        if !config[:resource].nil? then
            case config[:resource]
            when 'images'
                filetypes << 'gif'
                filetypes << 'jpg'
                filetypes << 'jpeg'
                filetypes << 'png'
                filetypes << 'svg'
            when 'javascript'
                filetypes << 'js'
                filetypes << 'coffee'
            when 'styles'
                filetypes << 'css'
                filetypes << 'sass'
                filetypes << 'less'
            end
        end

        if filetypes.empty?
            raise OptionParser::MissingArgument
        else
            puts "\e[0;93mHostname:\e[0m #{config[:hostname]}"
            puts "\e[0;93mTarget:\e[0m #{config[:target]}"
            puts "\e[0;93mScan-by:\e[0m #{config[:scanby]}"
            puts "\e[0;93mExtensions:\e[0m #{filetypes.join(', ')}"
            puts '---------------'
            response = open(
                config[:target],
                'User-Agent'=>random_agent,
                'Referer'=>config[:hostname]
            )
            content = response.read

            puts "\e[0;93mResources found:\e[0m"
            if filetypes.empty? then
                puts "Filetypes empty."
            else
                lines = content.gsub("\"\n",'').gsub("\n",'').split('>')
                lines.each do |line|
                    if match = line.match(/<base href=.([a-zA-Z0-9:\.\-\/_ ]+)./) then
                        puts "\e[0;93mBase path:\e[0m #{match[1]}"
                    else
                        if !line.empty? then
                            line = "#{line.chomp}>"
                            regex = //
                            if config[:scanby] == 'tag' then
                                case config[:resource]
                                when 'images'
                                    regex = /<img.*src=.([a-zA-Z0-9=:\?\.\-\/_ ]+)..*>/
                                when 'javascript'
                                    regex = /<script.*src=.([a-zA-Z0-9=:\?\.\-\/_ ]+)..*>/
                                when 'styles'
                                    regex = /<link.*href=.([a-zA-Z0-9=:\?\.\-\/_ ]+)..*>/
                                end
                                if match = line.match(regex) then
                                    puts "  #{match[1]}"
                                end
                            elsif config[:scanby] == 'extension' then
                                filetypes.each do |filetype|
                                    if match = line.match(/(href|src)=.([a-zA-Z0-9=:\?\.\-\/_ ]+\.#{filetype}).( |>)/) then
                                        puts "  #{match[2]}"
                                    end
                                end
                            end
                        end
                    end
                end
            end
        end
    end
rescue OptionParser::MissingArgument, OptionParser::InvalidOption
    puts optparse
    puts
    puts "\e[0;91mError.\e[0m Missing argument."
    # puts $!.to_s
end

File: remoteassets.go
/**
 * Remote Assets
 * http://cixtor.com/
 * https://github.com/cixtor/mamutools
 * http://en.wikipedia.org/wiki/Digital_asset
 *
 * List all the resources (assets) loaded in a specific page on a website.
 *
 * A digital asset is any item of text or media that has been formatted into a
 * binary source that includes the right to use it. A digital file without the
 * right to use it is not an asset. Digital assets are categorised in three major
 * groups which may be defined as textual content (digital assets), images (media
 * assets) and multimedia (media assets).
 */

package main

import (
    "os"
    "fmt"
    "flag"
    "net/url"
    "regexp"
    "strings"
    "net/http"
    "io/ioutil"
)

var remote_loc = flag.String("url", "", "Specify the remote location to scan")
var filetype = flag.String("filetype", "", "Specify the filetype to look in the remote location specified")
var resource = flag.String("resource", "", "Specify a group of filestypes to include in the scanning: images, javascript, styles")
var get_all = flag.Bool("all", false, "Show all the resources found in the site")
var colored = flag.Bool("colored", false, "Whether the information must be highlighted or not")
var verbose = flag.Bool("verbose", false, "Whether the request details must be printed or not")
var unique = flag.Bool("unique", false, "Whether the duplicated files will be counted or not")

func fail(message string) {
    flag.Usage()
    fmt.Printf("\nError: %s\n", message)
    os.Exit(1)
}

func print_colored(var_title string, var_value string) {
    if *colored {
        fmt.Printf( "\033[1;34m%s\033[0m: %s\n", var_title, var_value )
    } else {
        fmt.Printf( "%s: %s\n", var_title, var_value )
    }
}

func main() {
    flag.Usage = func(){
        fmt.Println("Remote Assets")
        fmt.Println("  http://cixtor.com/")
        fmt.Println("  https://github.com/cixtor/mamutools")
        fmt.Println("  http://en.wikipedia.org/wiki/Digital_asset")
        fmt.Println("Usage:")
        flag.PrintDefaults()
    }

    flag.Parse()

    *remote_loc = strings.TrimSpace(*remote_loc)
    if *remote_loc == "" {
        fail("Remote location not specified")
    }

    re := regexp.MustCompile(`^(http|https):\/\/`)
    var url_scheme []string = re.FindStringSubmatch(*remote_loc)
    if url_scheme == nil {
        *remote_loc = fmt.Sprintf("http://%s", *remote_loc)
    }

    location, err := url.Parse(*remote_loc)
    if err != nil {
        fail("URL malformation detected")
    }

    var filetypes []string
    var filetype_options = map[string][]string {
        "images": []string{ "gif", "jpg", "jpeg", "png", "svg" },
        "javascript": []string{ "js", "coffee" },
        "styles": []string{ "css", "sass", "less" },
    }

    if *get_all == true {
        for _, exts := range(filetype_options) {
            for _, ext := range(exts) {
                filetypes = append(filetypes, ext)
            }
        }
    } else if *filetype != "" {
        filetypes = append(filetypes, *filetype)
    } else if *resource != "" {
        switch *resource {
        case "images":
            filetypes = filetype_options[*resource]
        case "javascript":
            filetypes = filetype_options[*resource]
        case "styles":
            filetypes = filetype_options[*resource]
        }
    }

    if len(filetypes) == 0 {
        fail("Filetype list is empty")
    }

    var indentation string = ""
    if *verbose {
        indentation = "- "
        print_colored( "Hostname", location.Host )
        print_colored( "Target", *remote_loc )
        print_colored( "Extensions", strings.Join(filetypes, ", ") )
    }

    resp, err := http.Get(*remote_loc)
    if err != nil {
        fail("The site could not be scanned")
    }

    defer resp.Body.Close()
    body, err := ioutil.ReadAll(resp.Body)

    var find_base_path bool = true
    var content_clean string = strings.Replace( string(body), "\n", "", -1 )
    var lines []string = strings.Split(content_clean, ">")
    var unique_files = make(map[string]int)

    var file_ext_str string = strings.Join(filetypes, "|")
    file_ext_re := regexp.MustCompile(`(href|src)=['"]([a-zA-Z0-9=:\?\.\-\/_ ]+)\.(`+file_ext_str+`)['"]`)
    base_re := regexp.MustCompile(`<base href=['"]([a-zA-Z0-9:\.\-\/_ ]+)['"]`)

    for _, line := range(lines) {
        line = line + ">"

        if find_base_path && *verbose {
            var base_match []string = base_re.FindStringSubmatch(line)
            if base_match != nil {
                find_base_path = false
                print_colored( "Base path", base_match[1] )
                continue
            }
        }

        var file_match []string = file_ext_re.FindStringSubmatch(line)
        if file_match != nil {
            var file_path string = file_match[2] + "." + file_match[3]
            if count, ok := unique_files[file_path]; ok {
                unique_files[file_path] = count + 1
            } else {
                unique_files[file_path] = 1
            }

            if *unique != true {
                fmt.Printf("%s%s\n", indentation, file_path)
            }
        }
    }

    if *unique {
        for file_url, file_count := range(unique_files) {
            fmt.Printf( "(%d) %s\n", file_count, file_url )
        }
    }
}
